Implements JSON-RPC as a new message format for the Node methods and events

joelalejandro opened the pull request:
Resolves #1614.

While tests are passing in all affected packages, please review this PR with 🦅 eyes, since the scope is quite big and the surface area, albeit not implies altering core functionality (such as controller logic), does modify the way the Node talks to the outer world.

Basically every NodeMessage is now wrapped into a JSONRPC payload, as described in the master issue for this PR. Node-to-node events were left out after discussing with @snario, as there's a plan to rethink that aspect of the messaging layer.

ebryn commented on packages/node/rollup.config.js:
```diff
@@ -63,6 +64,13 @@ export default [
+          "../../node_modules/rpc-server/dist/index.js": [
+            "jsonRpcMethod",
+            "jsonApiOperation",
+            "jsonApiType",
```
the `jsonApi*` exports can probably be omitted?

joelalejandro replied:
Sure thing.

cf19drofxots commented:
@joelalejandro note that https://github.com/counterfactual/monorepo/pull/1618 exposes [one more method](https://github.com/counterfactual/monorepo/pull/1618/files#diff-a76195ae40c1f3b11462137d6b5acb66R276) that'll need to be consolidated with this new message format.

joelalejandro commented:
@cf19drofxots 012ebf08 exposes `chan_getProposedAppInstances`.

cf19drofxots commented on packages/node/src/request-handler.ts:
```diff
@@ -72,7 +78,8 @@ export class RequestHandler {
           result: await this.methods.get(methodName)(this, req.params)
         };
-        this.outgoing.emit(req.type, res);
+        // this.outgoing.emit(req.type, res);
```
Stale?

joelalejandro replied:
Yeah

cf19drofxots commented on packages/node/src/node.ts:
```diff
@@ -49,6 +51,7 @@ export class Node {
   // initialized in the `asynchronouslySetupUsingRemoteServices` function
   private signer!: HDNode;
   protected requestHandler!: RequestHandler;
+  public router: NodeRouter = {} as NodeRouter;
```
Shouldn't this be private?

joelalejandro replied:
Once we reimplement `Node.call` to use the JSONRPC `router.dispatch()`, we can make it private/protected.

cf19drofxots commented on packages/playground-server/src/node.ts:
```diff
@@ -257,22 +256,27 @@ export class NodeWrapper {
 }
 
 export async function onDepositConfirmed(response: DepositConfirmationMessage) {
+  console.log("onDepositConfirmed hit");
```
Stale debug line

joelalejandro replied:
Yeap

cf19drofxots commented on packages/playground-server/src/node.ts:
```diff
@@ -290,11 +294,20 @@ export async function onDepositConfirmed(response: DepositConfirmationMessage) {
-      type: NodeTypes.MethodName.DEPOSIT,
-      params: response.data as NodeTypes.DepositParams
-    });
+    // await NodeWrapper.getInstance().call(NodeTypes.MethodName.DEPOSIT, {
```
Stale

joelalejandro replied:
Also, yes

joelalejandro commented on packages/cf-wallet.js/src/provider.ts:
```diff
@@ -7,10 +7,21 @@ import {
 import { CounterfactualEvent, EventType } from "./types";
 
+export const jsonRpcMethodNames = {
+  [Node.MethodName.GET_FREE_BALANCE_STATE]: "chan_getFreeBalanceState",
```
@cf19drofxots A standard method naming for the JSONRPC API for the Node will be implemented once related discussions are resolved. @ebryn was looking into that, if I recall correctly.

joelalejandro commented on packages/cf-wallet.js/test/provider.spec.ts:
```diff
@@ -35,61 +39,40 @@ describe("CF.js Provider", () => {
 
   it("throws generic errors coming from Node", async () => {
-    expect.assertions(2);
+    // expect.assertions(2);
```
Yeap, this should be uncommented.

joelalejandro commented on packages/cf-wallet.js/test/provider.spec.ts:
```diff
@@ -35,61 +39,40 @@ describe("CF.js Provider", () => {
 
   it("emits an error event for orphaned responses", async () => {
-    expect.assertions(2);
+    // expect.assertions(2);
```
Yeap, this should be uncommented.

joelalejandro commented on packages/cf-wallet.js/test/provider.spec.ts:
```diff
@@ -35,61 +39,40 @@ describe("CF.js Provider", () => {
       await provider.getFreeBalanceState("foo");
     } catch (e) {
-      expect(e.data.errorName).toBe("unexpected_message_type");
+      expect(e.result.data.message).toBe("Music too loud");
```
Don't shoot the refactorer, I'm not the author of this 💎  :P

joelalejandro commented on packages/cf-wallet.js/test/provider.spec.ts:
```diff
@@ -132,26 +120,31 @@ describe("CF.js Provider", () => {
 
     it("can install an app instance virtually", async () => {
-      expect.assertions(7);
+      // expect.assertions(7);
```
Yeap, this should be uncommented.

joelalejandro commented on packages/cf.js/rollup.config.js:
```diff
@@ -30,14 +31,10 @@ export default {
     nodeResolve({
       only: [...bundledDependencies]
     }),
-    commonjs({
```
Wasn't relevant to the scope of the PR. This kind of iterations can be done in smaller, specifically-focused PRs.

joelalejandro commented on packages/cf.js/src/app-instance.ts:
```diff
@@ -86,6 +86,7 @@ export class AppInstance {
       }
     );
     const result = response.result as Node.GetStateResult;
+    console.log("Result for getState", result);
```
Yeap, this should be gone

joelalejandro commented on packages/cf.js/test/app-instance.spec.ts:
```diff
@@ -38,15 +39,21 @@ describe("CF.js AppInstance", () => {
       nodeProvider.onMethodRequest(Node.MethodName.GET_STATE, request => {
-        expect(request.type).toBe(Node.MethodName.GET_STATE);
-        const params = request.params as Node.GetStateParams;
+        console.log("Callback for GET_STATE executed");
```
Also should be gone

snario commented on packages/cf-wallet.js/test/fixture.ts:
```diff
@@ -22,19 +24,22 @@ export class TestNodeProvider implements INodeProvider {
   }
 
-  public simulateMessageFromNode(message: Node.Message) {
+  public simulateMessageFromNode(message: any) {
```
There seem to be many uses of `any` like this; are there no type enforcements to be made?

joelalejandro replied:
Yeap, working on them.

joelalejandro commented on packages/node/src/methods/app-instance/install-virtual/controller.ts:
```diff
@@ -12,6 +13,9 @@ import { installVirtual } from "./operation";
 export default class InstallVirtualController extends NodeController {
   public static readonly methodName = Node.MethodName.INSTALL_VIRTUAL;
 
+  @jsonRpcMethod("chan_installVirtual")
```
This was implemented by @snario. Some of the controllers require a queue. If I decorated `executeMethodImplementation` directly, the queue management wouldn't be invoked.

snario replied:
The other ones don't use `enqueueByShard`

snario commented on packages/cf.js/src/provider.ts:
```diff
@@ -16,6 +17,21 @@ import {
   UpdateStateEventData
 } from "./types";
 
+export const jsonRpcMethodNames = {
```
Why do we even need this mapping? Why isn't each key actually defined as the value in the `Node.MethodName` enum?

joelalejandro replied:
I didn't want to alter any of the current messaging layer. There's actually en enum called `Node.RpcMethodName` to leave the original one untouched. If you prefer to replace the old names with the new ones, just let me know.

joelalejandro commented on packages/cf-wallet.js/src/provider.ts:
```diff
@@ -249,34 +260,57 @@ export class Provider {
     params: Node.MethodParams
-  ): Promise<Node.MethodResponse> {
-    const requestId = new Date().valueOf().toString();
+  ): Promise<any> {
```
Yes, for now it can still be `Node.MethodResponse`.

joelalejandro commented on packages/cf-wallet.js/src/provider.ts:
```diff
@@ -249,34 +260,57 @@ export class Provider {
+          method: jsonRpcMethodNames[methodName],
+          id: requestId
+        });
+      } else {
```
This was a safe-guard to compare functionality between a migrated-to-JSONRPC method and a NodeMessage-based method. It shouldn't be needed anymore.

joelalejandro commented:
@cf19drofxots @snario I think I've addressed all of your concerns regarding the PR. What should we do with versioning? This PR impacts the following packages:

- `cf.js`
- `cf-wallet.js`
- `node`
- `types`

I'm thinking a minor release bump for the first three packages, and a patch for `types`.

cf19drofxots commented:
That version updating makes sense.

snario commented on packages/cf.js/src/provider.ts:
```diff
@@ -2,10 +2,17 @@ import {
+  JsonRpcNotification,
+  JsonRpcResponse,
+  jsonRpcSerializeAsResponse
+  // jsonRpcSerializeAsNotification
```
//

cf19drofxots commented:
As soon as CI is green with this branch we can merge it.
